src/main/java/hexgui/sgf/SgfReader.java

```diff
@@ -0,0 +1,411 @@
+
+      if (name.equals("W")) {
+        HexPoint point = parseMove(val);
+        node.setMove(new Move(point, HexColor.WHITE));
+      } else if (name.equals("B")) {
+        HexPoint point = parseMove(val);
+        node.setMove(new Move(point, HexColor.BLACK));
+      } else if (name.equals("AB")) {
+        node.addSetup(HexColor.BLACK, parsePoint(val));
+      } else if (name.equals("AW")) {
+        node.addSetup(HexColor.WHITE, parsePoint(val));
+      } else if (name.equals("AE")) {
+        node.addSetup(HexColor.EMPTY, parsePoint(val));
+      } else if (name.equals("LB")) {
+        node.addLabel(val);
+      } else if (name.equals("FF")) {
+        node.setSgfProperty(name, val);
+        x = parseInt(val);
+        if (x < 1 || x > 4) throw sgfError("Invalid SGF Version! (" + x + ")");
+      } else if (name.equals("GM")) {
+        node.setSgfProperty(name, val);
+        if (!isroot) sgfWarning("GM property in non-root node!");
+        if (parseInt(val) != GM_HEXGAME) throw sgfError("Not a Hex game!");
+      } else if (name.equals("SZ")) {
+        node.setSgfProperty(name, val);
+        if (!isroot) sgfWarning("GM property in non-root node!");
+        Dimension dim = new Dimension();
+        String sp[] = val.split(":");
+        if (sp.length == 1) {
+          x = parseInt(sp[0]);
+          dim.setSize(x, x);
+        } else if (sp.length == 2) {
+          x = parseInt(sp[0]);
+          y = parseInt(sp[1]);
+          dim.setSize(x, y);
+        } else {
+          throw sgfError("Malformed boardsize!");
+        }
+        m_gameinfo.setBoardSize(dim);
+      } else if (name.equals("AP")) {
+        node.setSgfProperty(name, val);
+        String regex = "HexGui:0\\.[0-9](\\z|[^0-9].*)";
+        if (val.matches(regex)) {
+          // version HexGui:0.9 or earlier
+          m_swap_bug = true;
+        }
+      } else {
+        node.setSgfProperty(name, val);
+      }
+    }
+  }
+
+  // Parse an SGF "SimpleText" property value.
+  private String parseValue() throws SgfError, IOException {
+    int ttype = m_tokenizer.nextToken();
+    if (ttype != '[') {
+      throw sgfError("Property missing opening '['.");
+    }
+
+    StringBuilder sb = new StringBuilder(256);
+    boolean quoted = false;
+    while (true) {
+      int ch = m_reader.read();
+      if (ch < 0) {
+        throw sgfError("Property runs to EOF.");
+      }
+      // Don't rely on the default conversion, because
+      // sb.append(ch) would convert the integer to a string
+      // rather than a character.
+      char c = (char) ch;
+
+      if (!quoted) {
+        if (c == ']') {
+          break;
+        }
+        if (c == '\\') {
+          quoted = true;
+        } else {
+          if (Character.isWhitespace(c)) {
+            // The spec says "Whitespaces other than space
+            // must be converted to space".
+            sb.append(' ');
+          } else {
+            sb.append(c);
+          }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 291:301, 356:371.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.